### PR TITLE
[Reassociate] Preserve `nuw` and `nsw` on `mul` chains

### PR DESCRIPTION
**Pull request by goldsteinn**
- **[Reassociate] Add tests for preserving `nuw` and `nsw` on `mul` chains; NFC**
- **[Reassociate] Preserve `nuw` and `nsw` on `mul` chains**

Basically the same rules as `add` but we also need to ensure all
operands a non-zero.

Proofs: https://alive2.llvm.org/ce/z/jzsYht

### CODE REVIEW

File: llvm/lib/Transforms/Scalar/Reassociate.cpp [hunk 1]
```diff
@@ -725,7 +735,9 @@ void ReassociatePass::RewriteExprTree(BinaryOperator *I,
           // Note that it doesn't hold for mul if one of the operands is zero.
           // TODO: We can preserve NUW flag if we prove that all mul operands
           // are non-zero.
```
Comment (nikic) [inline llvm/lib/Transforms/Scalar/Reassociate.cpp:737]: TODO resolved.

File: llvm/test/Transforms/Reassociate/reassoc-mul-nuw.ll [hunk 2]
```diff
@@ -32,3 +32,157 @@ define i4 @nuw_preserve_non_zero(i4 %a, i4 %b, i4 %c) {
...
+; CHECK-NEXT:    ret i4 [[MUL_C]]
+;
+  %x0_nz = add nuw i4 %xx0, 1
+  %x1_nz = add nuw i4 %xx1, 1
+  %x2_nz = add nuw i4 %xx2, 1
+  %x3_nz = add nuw i4 %xx3, 1
+  %x0 = call i4 @llvm.umin.i4(i4 %x0_nz, i4 7)
+  %x1 = call i4 @llvm.umin.i4(i4 %x1_nz, i4 7)
+  %x2 = call i4 @llvm.umin.i4(i4 %x2_nz, i4 7)
+  %x3 = call i4 @llvm.umin.i4(i4 %x3_nz, i4 7)
```
Comment (nikic) [inline llvm/test/Transforms/Reassociate/reassoc-mul-nuw.ll:103]: I think it would be more obvious to write this as smax 0.

**Review by nikic — APPROVED** [review]
LGTM

File: llvm/lib/Transforms/Scalar/Reassociate.cpp [hunk 3]
```diff
@@ -539,6 +539,16 @@ static bool LinearizeExprTree(Instruction *I,
     Ops.push_back(std::make_pair(V, Weight));
     if (Opcode == Instruction::Add && Flags.AllKnownNonNegative && Flags.HasNSW)
       Flags.AllKnownNonNegative &= isKnownNonNegative(V, SimplifyQuery(DL));
+    else if (Opcode == Instruction::Mul) {
+      // To preserve NUW we need all inputs non-zero.
+      // To preserve NSW we need all inputs strictly positive.
```
Comment (akshayrdeodhar) [inline llvm/lib/Transforms/Scalar/Reassociate.cpp:544]: Similar to AllKnownNonNegative- its possible for AllKnownNonZero to be `true` in a case where an operand is 0, but `HasNUW` and `HasNSW` are false. Perhaps update the comment on line 71?

nbd, upto you.

**Review by akshayrdeodhar — APPROVED** [review]
LGTM

**Pull request closed without merging**